library/src/main/java/org/apache/apex/malhar/lib/window/impl/SpillableWindowedKeyedStorage.java

```diff
@@ -0,0 +1,136 @@
+package org.apache.apex.malhar.lib.window.impl;
+
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.concurrent.Immutable;
+import javax.validation.constraints.NotNull;
+
+import org.apache.apex.malhar.lib.state.spillable.SpillableByteArrayListMultimapImpl;
+import org.apache.apex.malhar.lib.state.spillable.SpillableByteMapImpl;
+import org.apache.apex.malhar.lib.state.spillable.managed.ManagedStateSpillableStateStore;
+import org.apache.apex.malhar.lib.utils.serde.Serde;
+import org.apache.apex.malhar.lib.window.Window;
+import org.apache.apex.malhar.lib.window.WindowedStorage;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+import com.datatorrent.netlet.util.Slice;
+
+/**
+ * Created by david on 7/15/16.
+ */
+public class SpillableWindowedKeyedStorage<K, V> implements WindowedStorage.WindowedKeyedStorage<K, V>
+{
+  @NotNull
+  private final ManagedStateSpillableStateStore store;
+
+  protected final SpillableByteMapImpl<ImmutablePair<Window, K>, V> internValues;
+  protected final SpillableByteArrayListMultimapImpl<Window, K> internKeys;
+
+  private SpillableWindowedKeyedStorage()
+  {
+    // for kryo
+    store = null;
+    internValues = null;
+    internKeys = null;
+  }
+
+  public SpillableWindowedKeyedStorage(long bucket, byte[] identifier, Serde<Window, Slice> serdeWindow, Serde<K, Slice> serdeKey, Serde<ImmutablePair<Window, K>, Slice> serdeWindowKey, Serde<V, Slice> serdeValue)
+  {
+    store = new ManagedStateSpillableStateStore();
+    store.getCheckpointManager().setNeedBucketFile(false);
+    internValues = new SpillableByteMapImpl<>(store, identifier, bucket, serdeWindowKey, serdeValue);
+    internKeys = new SpillableByteArrayListMultimapImpl<>(store, identifier, bucket, serdeWindow, serdeKey);
+  }
+
+  @Override
+  public boolean containsWindow(Window window)
+  {
+    return internKeys.containsKey(window);
+  }
+
+  @Override
+  public long size()
+  {
+    return internKeys.size();
+  }
+
+  @Override
+  public void remove(Window window)
+  {
+    List<K> keys = internKeys.get(window);
+    for (K key : keys) {
+      internValues.remove(new ImmutablePair<>(window, key));
+    }
+    internKeys.removeAll(window);
+  }
+
+  @Override
+  public void migrateWindow(Window fromWindow, Window toWindow)
+  {
+    List<K> keys = internKeys.get(fromWindow);
+    internValues.remove(toWindow);
+    for (K key : keys) {
+      internKeys.put(toWindow, key);
+      ImmutablePair<Window, K> oldKey = new ImmutablePair<>(fromWindow, key);
+      ImmutablePair<Window, K> newKey = new ImmutablePair<>(toWindow, key);
+
+      V value = internValues.get(oldKey);
+      internValues.remove(oldKey);
+      internValues.put(newKey, value);
+    }
+    internKeys.removeAll(fromWindow);
+  }
+
```

Comment:
Will do. Can you explain a little bit what bucket and identifier actually do here? Is it something that is used to name the directories in HDFS eventually?